test/e2e/autoscale_test.go

```diff
@@ -357,12 +385,152 @@ func assertAutoscaleUpToNumPods(ctx *testContext, curPods, targetPods float64, d
 	}
 }
 
+func getUniqueHostPerConnection(t *testing.T, ctx *testContext, size int) (map[string]*websocket.Conn, error) {
+	clients := Setup(t)
+	uniqueHostConns := make(map[string]*websocket.Conn, size)
+
+	for {
+		connLen := len(uniqueHostConns)
+		if connLen >= size {
+			break
+		}
+
+		mux := sync.Mutex{}
+		gr := errgroup.Group{}
+		for i := 0; i < size; i++ {
+			gr.Go(func() error {
+				conn, err := connect(t, clients, ctx.names.URL.Hostname())
+				if err != nil {
+					return err
+				}
+
+				if err = conn.WriteMessage(websocket.TextMessage, []byte("ping")); err != nil {
+					return err
+				}
+
+				_, recv, err := conn.ReadMessage()
+				if err != nil {
+					return err
+				}
+				host := string(recv)
+
+				mux.Lock()
+				defer mux.Unlock()
+				if _, ok := uniqueHostConns[host]; !ok {
+					uniqueHostConns[host] = conn
+				} else {
+					conn.Close()
+				}
+				return nil
+			})
+		}
```

Comment:
This will generate more than `size`  entries, given proper situation.
E.g. you want size=3, you start 3 goroutines, you get 2 unique hosts. Now you start another 3 gorouties, and get another 2 unique ones. Now you have 4...

Reply:
yeah but calculating the diff between the actual and the expected count on every cycle was a bit ugly. also current implementation doesn't impact the behavior of the test. since we only spin up 4 pods, most of the time I get 1 or 2 extra pods which isn't a big deal. this seemed simpler and cleaner, so i left it as is. If you feel strongly about the level of determinism this needs to have, I will take another pass.

Reply:
yes, but what you can do instead, is that each of the go routines that you spun up, has a loop inside , that does it until `LoadOrStore` returns true, i.e. you move your loop from outside, to the inside. This will guarantee exactly `size` results in the end. Obvious if you have at least `size` pods :)

Reply:
yup that sounds good! :)